4081 Get latest cases from iquery pages

albertisfu opened the pull request:
This PR introduces two mechanisms, as described in #4081, to keep the latest cases from iQuery pages up to date.

The first mechanism is the `iquery_pages_probing_daemon`, which iterates over all the `district_or_bankruptcy_pacer_courts` excluding `["uscfc", "arb", "cit"]` and schedules `iquery_pages_probing` tasks for each court on every iteration.

Between each court, it will wait for `IQUERY_PROBE_WAIT / len(court_ids)` by default, which is 5 minutes. This means that approximately every 5 minutes, a new `iquery_pages_probing` task will be scheduled (only if the previous court task has already finished and workers can keep up with all the tasks).

### The `iquery_pages_probing` Task

The task works as follows:

1. It retrieves the current court `iquery_pacer_case_id_final` stored in Redis.
2. It performs up to `IQUERY_PROBE_ITERATIONS` (default 10) iterations following a geometric binary sequence. For example, if the initial `iquery_pacer_case_id_final` is 0, the probing sequence will be: `1, 2, 4, 8, 16, 32, 64, 128, 256`.
3. The pattern will depend on the starting `pacer_case_id`. For example, if the starting ID is 1,000, the pattern will be: `1001, 1002, 1004, 1008, 1016, 1032, 1064, 1128, 1256`.
4. If no hits are found in the previous probing task (`court_probe_cycle_no_hits > 1`), a random 5% jitter will be added to the geometric binary sequence. For instance, if the last value in the sequence is 256, the jitter can be between 1-13 and will be appended to each value in the sequence. If the jitter is 6: `1001+6, 1002+6, 1004+6, 1008+6, 1016+6, 1032+6, 1064+6, 1128+6, 1256+6`.
5. The probing task will request the case query page for each `pacer_case_id` on PACER.
6. Every hit will be stored into CL.
7. If there are 2 consecutive empty `pacer_case_id`s in the sequence, the probing will be aborted.
8. Only the latest hit found will trigger a sweep signal, with the `from_iquery_scrape` parameter set to `false` in the last hit.
9. The `iquery.probing.enqueued` semaphore will be cleaned up when finishing the task so other probing tasks can be scheduled in the next iteration.

**Behavior on Common Errors**

The `iquery_pages_probing` task is designed to retry common errors for each probe instead of retrying the whole task from scratch.

- `query_iquery_page` is used to request a single case iQuery page and can be retried independently up to 3 times in case of a `Timeout` or `PacerLoginException`.
- If 3 retries fail on a case iQuery page due to a `Timeout`, the main task will trigger a wait of `IQUERY_COURT_TIMEOUT_WAIT` (default 10 seconds) before requesting the next `pacer_case_id`. If 3 `Timeouts` are raised at the task level, it will be aborted and a `court_wait` equal to `IQUERY_COURT_BLOCKED_WAIT` (default 10 minutes) will be set, so no other probing task for the court will be scheduled for 10 minutes.
- An `HTTPError` when requesting a case iQuery page means that a non-200 status code was returned, likely indicating a block from the court (I saw two status code returned by a court that blocked my IP 404 and  403). If this occurs, a `court_wait` equal to `IQUERY_COURT_BLOCKED_WAIT` will be set and the task will be aborted.

### The iQuery Sweep Signal: `handle_update_latest_case_id_and_schedule_iquery_sweep`

This is the second mechanism required to complete the probing-scraping process and works as follows:

1. It is aborted for dockets created by the sweep task itself or for non-last hits returned by the `iquery_pages_probing` task, as these dockets won't have a `pacer_case_id_final` greater than the one currently stored in Redis.
2. The task is also aborted if any of the following conditions are not met:
    - The docket was not just created.
    - The docket doesn't have a `pacer_case_id`.
    - The docket court doesn't belong to `Court.federal_courts.district_or_bankruptcy_pacer_courts().exclude(pk__in=["uscfc", "arb", "cit"])`.

If the docket meets the previous conditions, the `update_latest_case_id_and_schedule_iquery_sweep` method is called once the transaction is committed to avoid any errors in the following process that could prevent the docket from being properly saved or delay the save process.

**`update_latest_case_id_and_schedule_iquery_sweep`**

This method is executed after the docket transaction that triggered the signal is committed.

The method process is wrapped within an atomic Redis lock (`acquire_atomic_redis_lock`), which uses a Lua script to make other processes wait until the lock is released, avoiding race conditions when getting and updating `iquery_pacer_case_id_final` and `iquery_pacer_case_id_status`. This prevents duplicate processing of some `pacer_case_id`s that are in progress or already scheduled.

- The method checks if the docket `incoming_pacer_case_id` is greater than the current `iquery_pacer_case_id_final`. If so, the `iquery_pacer_case_id_final` is updated with the new value.
- The method will schedule a `make_docket_by_iquery` task for each `pacer_case_id` that needs to be scraped, which are the `pacer_case_ids` between `iquery_pacer_case_id_status` and the updated `iquery_pacer_case_id_final`.
- Each `make_docket_by_iquery` task is scheduled with a delay (countdown) of 1 second from the previous one to maintain the rate of requesting 1 case iQuery page per second per court.
- The maximum number of tasks that can be scheduled at once is controlled by `IQUERY_SWEEP_BATCH_SIZE` (default 10,800), which is half of the Celery `visibility_timeout` (21,600 seconds) to avoid a runaway of Celery tasks. If `IQUERY_SWEEP_BATCH_SIZE` is reached, a new batch will schedule the remaining tasks starting from a countdown of 0 seconds.
- Once all the tasks are scheduled, the `iquery_pacer_case_id_status` is updated with the latest `pacer_case_id` scheduled, and the atomic Redis lock is released (`release_atomic_redis_lock`).

### Notes, Questions, and Concerns

- Before deployment, we need to set the `iquery_pacer_case_id_status` for all courts. This could be set from the latest `pacer_case_id` stored from each court from the date the last whole iQuery scrape finished.
- One concern is that some courts could have accumulated thousands of cases since the last scrape date. For example, if a court has 20,000 cases, each probing task will trigger a sweep task retrieval to get up to 256 cases on each sweep task, slowly catching up with all the cases in PACER. However, this could be problematic if a user uploads one of the latest dockets in PACER for a court, potentially triggering a sweep task that schedules ~20,000 cases in batches of 10,800 tasks, which could overwhelm the workers and take too much time, especially if this happens for many courts simultaneously.
- One strategy to avoid this is to only trigger the sweep task for the latest hit saved by the probing daemon, ignoring user docket iQuery uploads until we catch up with PACER. Once it's up to date. The number of dockets created every 5 minutes in a court should be relatively low and we could enable the signals for everything.
- An alternative is to make the sweep task resilient enough to handle thousands of tasks scheduled at once, which would require evaluating some variables:
    - Will we use an independent queue with dedicated workers for these tasks? How many workers will we have? This would help us calculate the maximum requests we can handle per second and consider a celery throttle.
- Another concern is that scheduling thousands of tasks at once will require a longer expiration for the `acquire_atomic_redis_lock` so other threads can wait long enough until the schedule is completed. However, it's not ideal to make other processes wait too long for the lock to be released. If a pod with a sweep signal waiting for the lock to be released dies, the signal will be lost, and if it contained a `pacer_case_id` greater than the current one stored, it will take more time to catch up. What is the recommended maximum wait time for a thread regarding pod lifetime?


Let me know what do you think.

albertisfu commented:
I was testing `nmd` to retrieve dockets after `502527`. It retrieved some of them but got stuck after `502531` due to a gap in that region. The issue was that the approach aborted the probing after two consecutive blank probes, which was insufficient here. So, I made the maximum number of consecutive blank probes configurable by setting `IQUERY_PROBE_MAX_CONSECUTIVE_FAILURES`. I set the default to 5, which was enough to pass the gap in `nmd` and retrieve everything up to `502703`.

Also while checking this, I found something unusual, which I'm not sure is normal.

The probe detected that the latest `pacer_case_id` in `nmd` was `502703`, and it stored the docket with this `pacer_case_id`.

However, when the signal retrieved everything from the previous `pacer_case_id_status` to `502703`, it found the same docket for `502703` as `502702`. The docket was then updated and saved with `pacer_case_id` `502702`.

I reviewed the case in PACER and found that `502702` and `502703` seem to retrieve the same case, or at least the case name and docket number are the same:

![Screenshot 2024-06-05 at 11 53 15 a m](https://github.com/freelawproject/courtlistener/assets/486004/b965fd36-ea49-47df-8ee1-9b73ab81b0f1)

However they do have different URLS:

`502703` https://ecf.nmd.uscourts.gov/cgi-bin/iquery.pl?150547770585821-L_1_0-1
`502702` https://ecf.nmd.uscourts.gov/cgi-bin/iquery.pl?202169394224075-L_1_0-1

Inspecting the HTML, both shows `502702` 
![Screenshot 2024-06-05 at 11 54 51 a m](https://github.com/freelawproject/courtlistener/assets/486004/a44492ea-c12a-4c88-bcff-458a910f4193)

So, I am wondering if it's the same case or not. If it's not the same case, we need to change the match method to avoid overriding the case with a different `pacer_case_id`, as in this case, it was matched by the `docket_number`.

mlissner commented:
> I am wondering if it's the same case or not. If it's not the same case, we need to change the match method to avoid overriding the case with a different pacer_case_id, as in this case, it was matched by the docket_number

Yes, it's pretty normal for docket numbers to be the same on several criminal dockets due to the #2185. 

The matching algo should match on the pacer_case_id + docket number and then broaden to just the docket number, no?

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+    interval_step=5 * 60,
+    ignore_result=True,
+)
+def iquery_pages_probing(
```
```suggestion
def probe_iquery_pages(
```

albertisfu replied:
Changed.

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+    bind=True,
+    autoretry_for=(RedisConnectionError,),
+    max_retries=3,
+    interval_start=1 * 60,
+    interval_step=5 * 60,
```
Hm, won't another be scheduled in these five minutes, and might that be an issue? Maybe it's better just to let it die?

albertisfu replied:
Yeah, that's right! I just removed the retries from the `probe_iquery_pages` task.

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+            )
+            break
+
+        probe_iteration, pacer_case_id_to_lookup = compute_next_binary_probe(
```
It'd be simpler not to mess around with probe_iteration in here, no? Why not just increment that outside the function?

albertisfu replied:
Yes, this is better. Now `compute_next_binary_probe`  only returns `pacer_case_id_to_lookup`

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+            iquery_pacer_case_id_final,
+            probe_iteration,
+            court_probe_cycle_no_hits,
+            probe_limit,
```
I think we can remove this too if we just apply a uniform jitter. Then we can just do: 

```
pacer_case_id_to_look_up = compute_next_binary_probe(iquery_pacer_case_id_final, iteration)
```

albertisfu replied:
Yeah, the final method looks like this:

```
pacer_case_id_to_look_up = compute_next_binary_probe(
    highest_known_pacer_case_id, iteration, jitter
)
```

`jitter` is required to pass so it's applied the same within all the probes in the task.

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -1160,6 +1168,8 @@ def make_docket_by_iquery(
     :param tag_names: A list of strings that should be added to the docket as
     tags
     :param log_results_redis: Log results in redis for the ready mix project
+    :param from_iquery_scrape: Weather this method was invoked by the iquery
```
As I said before, I think this would be better as "trigger_signal" or something like that?

albertisfu replied:
Sure, changed as described in a previous comment.

mlissner commented on cl/corpus_importer/management/commands/iquery_pages_probing_daemon.py:
```diff
@@ -0,0 +1,97 @@
+        iterations_completed = 0
+        q = settings.CELERY_IQUERY_QUEUE
+        # Create a queue equal than the number of courts we're doing.
+        throttle = CeleryThrottle(queue_name=q, min_items=len(court_ids))
```
I'm not sure you need a throttle here, do you? I think if we're going slowly enough the queue should stay pretty empty?

albertisfu replied:
Yeah, this is correct the throttle is not required here since we're going to schedule a probe task every 5 minutes and also in case the previous task hasn't finished the semaphore `enqueue_iquery_probe` will prevent another task for the same court to be scheduled.

mlissner commented on cl/recap/mergers.py:
```diff
@@ -1670,6 +1669,7 @@ def save_iquery_to_docket(
     d: Docket,
     tag_names: Optional[List[str]],
     add_to_solr: bool = False,
+    from_iquery_scrape=False,
```
Again, I think we can rename this.

albertisfu replied:
Changed!

mlissner commented on cl/corpus_importer/signals.py:
```diff
@@ -0,0 +1,122 @@
+    update_latest_case_id_and_schedule_iquery_sweep
+    """
+
+    if hasattr(instance, "from_iquery_scrape") and instance.from_iquery_scrape:
```
```suggestion
    if getsattr(instance, "from_iquery_scrape", False):
```

albertisfu replied:
Changed.

mlissner commented on cl/corpus_importer/signals.py:
```diff
@@ -0,0 +1,122 @@
+    """
+
+    r = get_redis_interface("CACHE")
+    court_id = docket.court.pk
```
This will save a query:
```suggestion
    court_id = docket.court_id
```

albertisfu replied:
Done!

mlissner commented on cl/corpus_importer/signals.py:
```diff
@@ -0,0 +1,122 @@
+        r.hget("iquery_pacer_case_id_status", court_id) or 0
+    )
+    incoming_pacer_case_id = int(docket.pacer_case_id)
+    updated_pacer_case_id = False
```
How about:

```suggestion
    found_higher_case_id = False
```

albertisfu replied:
Sure, changed!

mlissner commented on cl/corpus_importer/signals.py:
```diff
@@ -0,0 +1,122 @@
+    if updated_pacer_case_id:
+        task_scheduled_countdown = 0
+
+        while iquery_pacer_case_id_status + 1 < incoming_pacer_case_id:
+            tasks_processed_in_this_batch = 0
+
+            # Schedule tasks in batches of IQUERY_SWEEP_BATCH_SIZE to avoid
+            # a celery runaway scheduling tasks with countdowns larger than
+            # the celery visibility_timeout.
+            while (
+                tasks_processed_in_this_batch
+                < settings.IQUERY_SWEEP_BATCH_SIZE
+                and iquery_pacer_case_id_status + 1 < incoming_pacer_case_id
+            ):
```
Let's remove the batch size protection and do like you say to prevent new dockets from triggering the signal. I think that's easier than having this complexity with us forever, right?

albertisfu replied:
Yeah, this is better now as it will slowly catch up with the courts. I removed and also added the setting `IQUERY_SWEEP_UPLOADS_SIGNAL_ENABLED`, set to `False` by default. Once we catch up with the courts, we can set it to `True`.

mlissner commented on cl/corpus_importer/signals.py:
```diff
@@ -0,0 +1,122 @@
+        r.hget("iquery_pacer_case_id_final", court_id) or 0
+    )
+    iquery_pacer_case_id_status = int(
+        r.hget("iquery_pacer_case_id_status", court_id) or 0
```
I think a better name for this one is `iquery_pacer_case_id_current` since it's not the status, it's the current value, right?

albertisfu replied:
Sure changed!

mlissner reviewed (changes requested):
OK, I wound up making quite a few suggestions to try to make the code easier to undertand, mostly. 

Some of the suggestions simplify things, some just rename variables. 

Did you consider also using the throttle decorator as a protection against too many tasks? 

I think we can remove the batch protection that you've got and do the solution you proposed that avoids triggering iquery signals except from probes until all our probes are caught up.

albertisfu commented on cl/corpus_importer/tasks.py:
```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+            # Restart consecutive_empty_probes
+            consecutive_empty_probes = 0
+        else:
+            consecutive_empty_probes += 1
+            if (
+                consecutive_empty_probes
+                >= settings.IQUERY_PROBE_MAX_CONSECUTIVE_FAILURES
+            ):
+                # Too many consecutive blank probes in a row.
+                # Terminate the probing loop.
+                break
```
Well, I was using a similar approach, but with the difference that it checked 2 missed probes before aborting.

1: Hit
2: Miss
4: Miss
Abort

The problem here is that this is not enough to pass some small gaps, especially if they are at the beginning of the next probe.

For example, in this case, it seems common for the next pacer case to be blank, causing the probe to be aborted.

It is also possible that 4 is still blank, but 8 can be a hit.

The more probes we perform, the less likely it is that the process will get stuck in the first blank items.

1: Hit
2: Miss
4: Miss
8: Hit

1: Hit
2: Miss
4: Miss
8: Miss
16: Hit

So, setting up a counter for missed hits allows us to pass these kinds of gaps more easily. 
Does it make sense?

mlissner replied:
I think so, but I'm suggesting that until you get a hit, you just go up to 256, but once you get a hit you abort as soon as you get a miss. 

So the algo is:
 - Probe until you get a hit.
 - Once you get a hit, keep probing until you get a miss.

In five minutes, you'll come back and do the same, which will get you past any gaps.

albertisfu replied:
> I'm suggesting that until you get a hit, you just go up to 256, but once you get a hit you abort as soon as you get a miss.

Great, just to confirm I understood correctly:

Every probe task will always have to perform at least 9 probes?
`1, 2, 4, 8, 16, 32, 64, 128, 256`

And if 256 is a hit, we should still probe 512, 1024, and so on until we get a miss and then abort?

Because the other interpretation of this algo, which is the problem I was describing, is:

Once you get a hit (let's say 1 is a hit), keep probing until you get a miss (2 is a miss). Probing is aborted on 2.
Which is not good.

mlissner replied:
Not quite. We'll go up to 256 every five minutes. We'll keep going higher if we get misses or hits, but once we get hits, we'll abort on the first subsequent miss.

> Once you get a hit (let's say 1 is a hit), keep probing until you get a miss (2 is a miss). Probing is aborted on 2.

That's OK. We'll get it on the next round.

albertisfu replied:
Got it! Sounds good now, I'll apply it.

albertisfu replied:
This algo is now implemented:

> So the algo is:
> 
> Probe until you get a hit.
> Once you get a hit, keep probing until you get a miss.

I tested it on a couple of courts and it worked smoothly!

albertisfu commented:
Thanks for your comments and suggestions; I'll be applying them. I just left a comment on the one regarding how to abort the probing, since it seems it's better to count the missed probes before aborting.

> The matching algo should match on the pacer_case_id + docket number and then broaden to just the docket number, no?

Yes, I reviewed the method, and this is the order it uses: first, it uses the `pacer_case_id + docket_number` if available, and then broadens the lookup if no match is found.

Here is the code for that part:
```
 if pacer_case_id:
        lookups = [
            {
                "pacer_case_id": pacer_case_id,
                "docket_number_core": docket_number_core,
            },
            {"pacer_case_id": pacer_case_id},
        ]
    if docket_number_core:
        lookups.extend(
            [
                {
                    "pacer_case_id": None,
                    "docket_number_core": docket_number_core,
                },
                {"docket_number_core": docket_number_core},
            ]
        )
```

So the lookup order is:

```
1.- {
    "pacer_case_id": pacer_case_id,
    "docket_number_core": docket_number_core,
}

2.- {"pacer_case_id": pacer_case_id}

3.- {
    "pacer_case_id": None,
    "docket_number_core": docket_number_core,
}

4.-  {"docket_number_core": docket_number_core}
```

In this case, the case was not matched by the first three lookups, but it was in the fourth.

So, I think we need to update the lookup to only be applied if there is no `pacer_case_id`:

Something like:

 ```
   if not pacer_case_id and docket_number_core:
        lookups.append(
                {"docket_number_core": docket_number_core},
        )
```

> I think we can remove the batch protection that you've got and do the solution you proposed that avoids triggering iquery signals except from probes until all our probes are caught up.

Great I'll remove the batch protection and I'll apply the required logic to control via a setting when to start hearing global signals once probes are caught up.


> Did you consider also using the throttle decorator as a protection against too many tasks?

Well, the problem with the throttle decorator is that it doesn't limit the number of tasks sent to the queue. It just reschedules them when the rate is above the limit. So, it won't solve the issue of overwhelming Redis with too many tasks.

I considered using the `CeleryThrottle` class within `update_latest_case_id_and_schedule_iquery_sweep` to control the tasks sent to the queue. 
Even though we'll initially apply the approach of slowly catching up with the courts, don't you think it could happen someday that many courts add many dockets within a 5-minute period? This could result in a queue with thousands of tasks. I think this situation may be rare and possibly manageable.

However, if we want to ensure we can handle that scenario, we could use `CeleryThrottle` with a `maybe_wait`. Using this approach, the process would have to wait long enough to schedule all the tasks, which means the lock would require a long expiration time. This can be problematic because if the pod handling the process dies, the lock won't be released until it expires, blocking other court processes until the lock expires.

If we decide to go this route, considering infrastructure limits, is there a recommended maximum process lifetime to avoid this issue? This way, I can consider a lock expiration equal to this duration and work backward. I think we could use `CeleryThrottle` to schedule the tasks in small batches. Each time we need to wait, we could release the previous lock, create a new one, and process the remaining tasks until completion.

mlissner commented:
> So, I think we need to update the lookup to only be applied if there is no pacer_case_id:

Makes sense.

>  don't you think it could happen someday that many courts add many dockets within a 5-minute period? 

Seems pretty unlikely, actually. I think adding new cases is a pretty manual thing. Even in that case, I think we'd just want to queue them up and let it process, so long as we don't overwhelm the court by hitting it too rapidly. 

When we've had memory problems in the past, that was usually millions of queued tasks.

With that in mind, do you think we need the throttling stuff or do you think we might be OK without it?

albertisfu commented:
> With that in mind, do you think we need the throttling stuff or do you think we might be OK without it?

Got it. Yeah, it sounds unlikely that once we catch up with the court to have a task overflow due to scheduling a huge number of tasks, so no throttling will be required to schedule tasks.

I'm thinking that the only scenario where we'd need throttling is to avoid exceeding the limit of 1 request per second per court.

Imagine the following scenario:

13:00:00
Current `pacer_case_id`: `100`
An upload says we should go up to `105`.
Five tasks are scheduled, 1 second apart from each other:

13:00:01 `101`
13:00:02 `102`
13:00:03 `103`
13:00:04 `104`
13:00:05 `105`


Then we receive another upload at 13:00:03 with a `pacer_case_id` of `110`.
That will schedule another five tasks:
13:00:04 `106`
13:00:05 `107`
13:00:06 `108`
13:00:07 `109`
13:00:08 `110`

That means at 13:00:04 and 13:00:05, the court rate of 1 request per second will be exceeded.

Here, I think the `@throttle` decorator can help, so tasks can be rescheduled for later if the rate is exceeded at a given moment.

mlissner commented:
Yep, let's do it then. Good point.

albertisfu commented on cl/tests/cases.py:
```diff
@@ -6,13 +6,17 @@
+)
 from cl.lib.redis_utils import get_redis_interface
-from cl.search.models import SEARCH_TYPES
+from cl.search.models import SEARCH_TYPES, Docket
```
Yeah, these were not required anymore, removed.

albertisfu commented on docker/django/docker-entrypoint.sh:
```diff
@@ -42,6 +42,9 @@ case "$1" in
 'sweep-indexer')
     exec /opt/courtlistener/manage.py sweep_indexer
     ;;
+'iquery_pages_probing_daemon')
```
Changed

albertisfu commented on cl/lib/redis_utils.py:
```diff
@@ -71,3 +73,67 @@ def delete_redis_semaphore(r: Union[str, Redis], key: str) -> None:
+    else
+        return 0
+    end
+    """
```
Yeah, you're right. Acquiring the lock doesn't require a Lua script. I thought the `NX` option (used to set the key only if it doesn't exist) wasn't available in the regular Python SET method, but it is. So this method can be simplified as follows:

```
identifier = str(uuid.uuid4())
    while True:
        if r.set(key, identifier, nx=True, px=ttl):
            return identifier
        time.sleep(0.1)
```
I've updated the method to avoid using the Lua script here.

For releasing the lock, the Lua script is still required since it performs two operations: `GET` and `DEL`

albertisfu commented on cl/corpus_importer/utils.py:
```diff
@@ -1024,3 +1025,44 @@ def get_court_id(raw_court: str) -> list[str]:
+def compute_next_binary_probe(
+    iquery_pacer_case_id_final: int,
+    probe_iteration: int,
+    court_probe_cycle_no_hits: int,
```
Yeah, applying  jitter to all from the first iteration simplifies this method, so `court_probe_cycle_no_hits` is no longer required. Instead, the `jitter` is passed to this method, which solves another issue I found. When the `jitter` was computed inside this method, each probe used a different `jitter`, which led to possible duplicate probes due to the jitter's randomness.

With this fix, the `jitter` is computed once within the task and used the same for all the probes.

albertisfu commented on cl/corpus_importer/utils.py:
```diff
@@ -1024,3 +1025,44 @@ def get_court_id(raw_court: str) -> list[str]:
+
+
+def compute_next_binary_probe(
+    iquery_pacer_case_id_final: int,
```
Thanks for this suggestion. Yeah, I think it's better: `highest_known_pacer_case_id`. The final key is `iquery:highest_known_pacer_case_id` because I added the `iquery:` prefix to all the related Redis keys in this for better organization.

albertisfu commented on cl/corpus_importer/management/commands/iquery_pages_probing_daemon.py:
```diff
@@ -0,0 +1,97 @@
+                    continue
+                throttle.maybe_wait()
+                try:
+                    newly_enqueued = enqueue_iquery_probing(court_id)
```
Changed.

albertisfu commented on cl/corpus_importer/management/commands/iquery_pages_probing_daemon.py:
```diff
@@ -0,0 +1,97 @@
+                    if newly_enqueued:
+                        # No other probing being conducted for the court.
+                        # Enqueue it.
+                        iquery_pages_probing.delay(court_id, testing)
```
Changed. Also updated the command name to `iquery_pages_probe_daemon`

albertisfu commented on cl/recap/mergers.py:
```diff
@@ -1736,3 +1739,40 @@ async def process_orphan_documents(
             pass
+
+
+@retry(IntegrityError, tries=3, delay=0.25, backoff=1)
```
Yeah, this method, `process_case_query_report`, is similar to `save_iquery_to_docket`. However, due to their different retry approaches, I thought it was better to create a new method.

`process_case_query_report` is executed within the probe task. If there is a race condition, an `IntegrityError` can be raised when saving the docket. In this case, the method for that docket is independently retried using the `@retry` decorator, instead of retrying the whole probe task, which could lead to re-processing previously stored cases .

The `save_iquery_to_docket` method is called by the `make_docket_by_iquery_base` task. Its retry method is based on Celery retries, so the entire task is retried.

albertisfu commented on cl/corpus_importer/tasks.py:
```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+
+
+@retry((requests.Timeout, PacerLoginException), tries=3, delay=0.25, backoff=1)
+def query_iquery_page(
```
Sure, in this case, I created a method called `query_case_query_report`, which encapsulates the common code so it can be reused by `query_iquery_page` and `make_docket_by_iquery_base`.

albertisfu commented on cl/corpus_importer/tasks.py:
```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+                "The court website is probably down. Waiting a bit before trying "
+                "the next pacer_case_id."
+            )
+            time.sleep(settings.IQUERY_COURT_TIMEOUT_WAIT)
```
Got it! Yeah, I've changed the approach here. So once a `Timeout` is raised within the task, it's just aborted, returning only the probes that were successful. The task is not scheduled for later. In this case, the daemon will schedule a new one in the next court iteration. Does that sound good?

albertisfu commented on cl/corpus_importer/tasks.py:
```diff
@@ -1215,7 +1225,167 @@ def make_docket_by_iquery(
+            court_id,
+            report_data[0],
+            report_data[1],
+            from_iquery_scrape=from_iquery_scrape,
```
I renamed it to `avoid_trigger_signal`, so by default, this is `True`. It is set to `False` only on the last hit found by the probing daemon.

albertisfu commented:
Thanks for all your comments and suggestions! I've applied them and left some comments above.

Here are some additional comments and notes:

- Added `make_docket_by_iquery_base`, a new method common to both `make_docket_by_iquery` and `make_docket_by_iquery_sweep`. The difference between the two methods is that the first doesn't use the `throttle_task` decorator to avoid Celery runaways in bulk importers, while `make_docket_by_iquery_sweep` uses it to control the court rate for the iquery sweep.
- The setting `IQUERY_SWEEP_UPLOADS_SIGNAL_ENABLED` will be disabled by default. Once we catch up with the courts, we should set it to `True`. This will prevent scheduling too many tasks in a short period of time due to new uploads.
- There is also a new setting: `IQUERY_SWEEP_COURT_RATE`, which defaults to `100/s`. This is required for tests to pass because it is not possible to override the `throttle_task` rate in tests. Therefore, it is set to `100/s` by default, and we should set it to `1/s` in the production environment variables.
- The related Celery tasks `probe_iquery_pages` and `make_docket_by_iquery_sweep` will use the same Celery queue defined by the setting `CELERY_IQUERY_QUEUE`. Let me know if they should go to a different queue.
- Also, before this is deployed, we'll need to set the `highest_known_pacer_case_id` and `pacer_case_id_current` for all the courts. To do this easily, I tweaked the `ready_mix_cases_project` command to set them:

`docker exec -it cl-django python /opt/courtlistener/manage.py ready_mix_cases_project --task set-case-ids --court-type all --date-filed 2024-06-12`

The `date_filed` should be changed to the date we finished the latest iquery scrape so we can ensure we'll get the latest `pacer_case_ids` for each court without gaps.

- Regarding the issue I found in `find_docket_object` lookups where dockets with the same `docket_number` but different IDs were overridden, I tweaked the lookups as follows:


```
    if docket_number_core and not pacer_case_id:
        # Sometimes we don't know how to make core docket numbers. If that's
        # the case, we will have a blank value for the field. We must not do
        # lookups by blank values. See: freelawproject/courtlistener#1531
        lookups.extend(
            [
                {
                    "pacer_case_id": None,
                    "docket_number_core": docket_number_core,
                },
                {"docket_number_core": docket_number_core},
            ]
        )
    elif docket_number and not pacer_case_id:
        # Finally, as a last resort, we can try the docket number. It might not
        # match b/c of punctuation or whatever, but we can try. Avoid lookups
        # by blank docket_number values.
        lookups.append(
            {"pacer_case_id": None, "docket_number": docket_number},
        )
```
I just wanted to confirm this behavior. The lookups where the `pacer_case_id` is `None` or not `provided` will be performed only on uploads where the `pacer_case_id` is not present, such as on dockets created by opinions. I tried to think of a scenario where this tweak could lead to duplicated dockets, but I believe that considering the combination of `pacer_case_id` and `docket_number` is unique in a court, it should not happen.

mlissner commented on cl/corpus_importer/signals.py:
```diff
@@ -0,0 +1,126 @@
+    """
+
+    r = get_redis_interface("CACHE")
+    court_id = str(docket.court_id)
```
court_id is a charfield.

albertisfu replied:
Yeah, I removed the `str` cast here. I had added it because pycharm was complaining about the court_id being an `int`, but I think it's just something weird from the editor that can't detect it as a character since mypy is not complaining about that.

mlissner reviewed (approved):
One tiny thing, but I think this looks great. The tweaks made a big difference in readability. 

This is complex enough that we should get some real eyes on it that aren't mine, so @ERosendo, if you can take a look, that'd be great. (Let me know if you feel like you've had too much PR review duty!)

ERosendo commented:
> if you can take a look, that'd be great

Sure! I'll take a look 👀 

> Let me know if you feel like you've had too much PR review duty!

I'm good. thanks checking in! 😃

ERosendo commented on cl/corpus_importer/tasks.py:
```diff
@@ -1203,21 +1220,228 @@ def make_docket_by_iquery(
+    court_id: str,
+    pacer_case_id: int,
+    using: str = "default",
+    tag_names: Optional[List[str]] = None,
```
We can refactor this as:

```suggestion
    tag_names: List[str] | None = None,
```

ERosendo commented on cl/corpus_importer/tasks.py:
```diff
@@ -1203,21 +1220,228 @@ def make_docket_by_iquery(
+    tag_names: Optional[List[str]] = None,
+    log_results_redis: bool = False,
+    avoid_trigger_signal: bool = False,
+) -> Optional[int]:
```
Same here:

```suggestion
) -> int | None:
```

ERosendo commented on cl/corpus_importer/tasks.py:
```diff
@@ -1203,21 +1220,228 @@ def make_docket_by_iquery(
+    court_id: str,
+    pacer_case_id: int,
+    using: str = "default",
+    tag_names: Optional[List[str]] = None,
```
```suggestion
    tag_names: List[str] | None = None,
```

ERosendo commented on cl/corpus_importer/tasks.py:
```diff
@@ -1203,21 +1220,228 @@ def make_docket_by_iquery(
+    tag_names: Optional[List[str]] = None,
+    log_results_redis: bool = False,
+    avoid_trigger_signal: bool = False,
+) -> Optional[int]:
```
```suggestion
) -> int | None:
```

ERosendo commented on cl/recap/mergers.py:
```diff
@@ -1670,6 +1668,7 @@ def save_iquery_to_docket(
     d: Docket,
     tag_names: Optional[List[str]],
     add_to_solr: bool = False,
+    avoid_trigger_signal=False,
```
We're missing the type hint for this parameter.

albertisfu replied:
Sure, added it.

ERosendo commented on cl/lib/redis_utils.py:
```diff
@@ -71,3 +73,54 @@ def delete_redis_semaphore(r: Union[str, Redis], key: str) -> None:
+        time.sleep(0.1)
+
+
+def release_redis_lock(r: Redis, key, identifier: str) -> int:
```
```suggestion
def release_redis_lock(r: Redis, key: str, identifier: str) -> int:
```

albertisfu replied:
Fixed.

ERosendo commented on cl/lib/redis_utils.py:
```diff
@@ -71,3 +73,54 @@ def delete_redis_semaphore(r: Union[str, Redis], key: str) -> None:
+    return f"update.pacer_case_id:{court_id}"
+
+
+def acquire_redis_lock(r: Redis, key, ttl: int) -> str:
```
```suggestion
def acquire_redis_lock(r: Redis, key: str, ttl: int) -> str:
```

albertisfu replied:
Fixed.

ERosendo commented on cl/corpus_importer/tasks.py:
```diff
@@ -1219,21 +1235,228 @@ def make_docket_by_iquery(
+        except HTTPError:
+            # Set expiration accordingly and value to 2 to difference from
+            # other waiting times.
+            r.setex(
+                f"iquery:court_wait:{court_id}",
+                settings.IQUERY_COURT_BLOCKED_WAIT,
+                2,
+            )
```
is there a reason to use `SETEX` over `SET`? [The documentation](https://redis.io/docs/latest/commands/setex/) for `SETEX` states that it's been deprecated since Redis version 2.6.12.  For setting a key with an expiration time, `SET` is the recommended alternative. 

```suggestion
            r.set(
                f"iquery:court_wait:{court_id}",
                2,
                ex = settings.IQUERY_COURT_BLOCKED_WAIT,
            )
```

albertisfu replied:
Yeah, there's no specific reason to use `setex` is the same to use just `set`. I changed it to set, thanks for noticing it.

ERosendo commented on cl/corpus_importer/signals.py:
```diff
@@ -0,0 +1,126 @@
+    if (
+        created
+        and instance.pacer_case_id
+        and getattr(instance, "court", None)
```
Is this check required? do we have any Docket records that haven't been linked to a specific court?

albertisfu replied:
Yeah, that check is not required. Court is a required relation. Also confirmed that in the dev DB there's no Docket with a court Null. Removed it.

ERosendo commented on cl/corpus_importer/utils.py:
```diff
@@ -1024,3 +1026,46 @@ def get_court_id(raw_court: str) -> list[str]:
+def compute_next_binary_probe(
+    highest_known_pacer_case_id: int, iteration: int, jitter: int
+) -> int:
+    """Compute the next binary  probe target for a given PACER case ID.
```
```suggestion
    """Compute the next binary probe target for a given PACER case ID.
```

ERosendo commented on cl/settings/project/corpus_importer.py:
```diff
@@ -0,0 +1,14 @@
+env = environ.FileAwareEnv()
+IQUERY_PROBE_ITERATIONS = env.int("IQUERY_PROBE_ITERATIONS", default=9)
+IQUERY_PROBE_WAIT = env.int("IQUERY_PROBE_WAIT", default=300)
+IQUERY_PROBE_MAX_CONSECUTIVE_FAILURES = env.int(
+    "IQUERY_PROBE_MAX_CONSECUTIVE_FAILURES", default=5
+)
```
I noticed this setting isn't referenced anywhere in the code.

albertisfu replied:
Yeah, this is unused now, removed.

ERosendo commented on cl/settings/project/corpus_importer.py:
```diff
@@ -0,0 +1,14 @@
+    "IQUERY_PROBE_MAX_CONSECUTIVE_FAILURES", default=5
+)
+IQUERY_COURT_BLOCKED_WAIT = env.int("IQUERY_COURT_BLOCKED_WAIT", default=600)
+IQUERY_COURT_TIMEOUT_WAIT = env.int("IQUERY_COURT_TIMEOUT_WAIT", default=10)
```
Similar to the previous setting, this one also appears unused. We can likely remove it for consistency.

albertisfu replied:
Thanks. yeah, this is unused; it was part of a previous architecture. Now we just abort the task on a timeout and wait for the next iteration.

ERosendo commented:
The code looks good. We can merge the PR after addressing the comments.

albertisfu commented:
Thanks @ERosendo I've applied your suggestions.

A couple of additional improvements I've applied:

- We need to prevent the daemon from running until the `highest_known_pacer_case_id` and `pacer_case_id_current` for each court are in place.

By running this command once the PR is merged, since the tweak in the command to update the keys in Redis is included in this PR:

```
docker exec -it cl-django python /opt/courtlistener/manage.py ready_mix_cases_project --task set-case-ids --court-type all --date-filed 2024-06-12
```

It is only necessary to confirm the `--date-filed` to the date the last full iquery scraper finished, ensuring there are no gaps preceding those `pacer_case_ids`.

To prevent the daemon from running, I added the `IQUERY_PROBE_DAEMON_ENABLED` default to `False`. Once the IDs are in place, it can be set to `True` and  restart the daemon.


- I added a logger into the daemon so we can monitor its running status and the scheduling of probes.
- Within `update_latest_case_id_and_schedule_iquery_sweep`, for safety, I added a check that computes the number of tasks to schedule. If there are more than 10,800 tasks (half of the max countdown according to the 6-hour Celery visibility_timeout), it will trigger an error and avoid scheduling the tasks. Although this might be rare, it's better to prevent it. This could occur if the court accidentally added a huge `pacer_case_id` instead of the next in sequence, or if we lost the `pacer_case_id_current` or `highest_known_pacer_case_id` in Redis. If this happens, we can investigate and take action without overwhelming Celery.
- In `probe_iquery_pages`, I added a try-except to catch `IntegrityError` in case the individual retries for `process_case_query_report` are exhausted. This might indicate an issue other than a race condition, allowing us to investigate the problem while still processing other reports in the probe.

Finally, confirming other settings that should be set as env vars:

Before merging it:
- `IQUERY_COURT_RATE` should be set to `1/s` from its default `100/s` for testing purposes.
- `IQUERY_COURT_BLOCKED_WAIT`: default to 600 seconds (10 minutes). Not sure if this duration is sufficient to handle a court block. Perhaps is too short?

After conditions are met:
- `IQUERY_PROBE_DAEMON_ENABLED`: default to `False`; once `pacer_case_ids` are set, it should be set to `True`.
- `IQUERY_SWEEP_UPLOADS_SIGNAL_ENABLED`: default to `False`; this will prevent uploads and other sources other than the probe daemon from scheduling an iquery scrape for a higher `pacer_case_id` that could overwhelm Celery. Initially, the probe daemon will slowly schedule iquery page scrapes until it catches up with the court so we could set it to `True` so that all other sources can also schedule iquery scrapes. Perhaps in a week or so, we can check if all the `pacer_case_ids` are mostly up to date with the court and enable it.

mlissner commented on cl/corpus_importer/management/commands/probe_iquery_pages_daemon.py:
```diff
@@ -0,0 +1,98 @@
+
+
+class Command(VerboseCommand):
+    help = "Run the iquery pages probing daemon."
```
```suggestion
    help = """Run the iquery pages probing daemon.

The goal of this daemon is to ensure that we always have every case in PACER.

It works by taking the highest pacer_case_id we know as of the last iteration,
and then doing geometric probing of higher numbers to discover the highest
current pacer_case_id in each court.

For example, if the highest ID we know about as of 12:00PT is 1000, we would
check ID 1002, then 1004, 1008, 1016, etc., until we stop finding valid cases.
Because cases can be sealed, deleted, etc, we also add jitter to the IDs we
select, to ensure that we don't get stuck due to bad luck.

Once the highest value is found, we schedule iquery download tasks one second
apart to fill in the missing section. For example, if we start at ID 1000, then
learn that ID 1032 is the highest, we'd create 31 celery tasks for items 1000
to 1032, and we'd schedule them over the next 31 seconds.

The last piece of this system is that we have content coming in from a lot of
sources all the time, so we use signals to backfill missing content as well.
For example, if we think 1,000 is the highest ID, and somebody RECAPs a docket
with ID of 1032, the signal will catch that and create tasks to fill in numbers
1001 to 1031.
"""
```

mlissner reviewed (approved):
I made one tweak to add a longer docstring. Otherwise, look good. Merging. Thanks for the discussion and tricky work on this. Eduardo, thanks for the review.

mlissner commented:
Alberto, I'm realizing I'm the blocker and the bottleneck for this. Do you think you could work with Ramiro to launch it? I don't think there's much left to do, IIRC, except for setting up a daemon and some variables, right?

albertisfu commented:
Of course. I'll work with Ramiro to get this done.

Just one question, we'll need to run this command first in order to update the latest `pacer_case_ids` we got from the courts without gaps.

`docker exec -it cl-django python /opt/courtlistener/manage.py ready_mix_cases_project --task set-case-ids --court-type all --date-filed 2024-05-29`

Do you remember the date the last iquery scrape finished? I remember it should be around May 29, but I also recall we got blocked from some courts and ran another scrape that should have finished a few days later.

albertisfu commented:
@blancoramiro  when you have a moment, please let me know so we can get this one launched.

First, we need to run the following script to set the initial `pacer_case_ids` to `1` for each PACER court on Redis:

```
from cl.lib.redis_utils import get_redis_interface
from cl.search.models import Court

court_ids = list(
            Court.federal_courts.district_or_bankruptcy_pacer_courts()
            .exclude(pk__in=["uscfc", "arb", "cit"])
            .values_list("pk", flat=True)
        )

r = get_redis_interface("CACHE")
for court_id in court_ids:
    r.hset("iquery:highest_known_pacer_case_id",court_id,1)
    r.hset("iquery:pacer_case_id_current", court_id, 1)

```

And then we can confirm IDS were properly set by:


```
from cl.lib.redis_utils import get_redis_interface

r = get_redis_interface("CACHE")
highest_known_pacer_case_id = r.hgetall("iquery:highest_known_pacer_case_id")
pacer_case_id_current = r.hgetall("iquery:pacer_case_id_current")

highest_known_pacer_case_id = {k: v for k, v in highest_known_pacer_case_id.items()}
pacer_case_id_current = {k: v for k, v in pacer_case_id_current.items()}

data = [
    {"court_id": court_id, "highest_known_pacer_case_id": highest_known_pacer_case_id[court_id], "pacer_case_id_current": pacer_case_id_current[court_id]}
    for court_id in highest_known_pacer_case_id.keys()
]
print(data)
```

The we need to add the following env vars:
`IQUERY_COURT_RATE` = `1/s`
`IQUERY_PROBE_DAEMON_ENABLED`= `True`

And finally, we need to deploy the daemon `probe-iquery-pages-daemon`, which is already in `docker-entrypoint.sh`.

The Celery queue where these tasks will be scheduled is defined in the environment variable:
`CELERY_IQUERY_QUEUE`.

The average number of tasks to be scheduled is around 200 every 5 minutes, so I think this task doesn't require its own workers.

Thank you!

albertisfu commented:
@mlissner  I confirmed that the steps needed to get the iquery scraper running are as described in the previous comment https://github.com/freelawproject/courtlistener/pull/4090#issuecomment-2223705681

Since we agreed to start scraping from `pacer_case_id` 1 for each PACER court, we need to set the required Redis keys for each court and `pacer_case_id:1`.

Then, set the provided environment variables:

- `IQUERY_COURT_RATE = 1/s`
- `IQUERY_PROBE_DAEMON_ENABLED = True`

Also, assign a queue for these tasks to `CELERY_IQUERY_QUEUE`.

Finally, deploy the daemon `probe-iquery-pages-daemon`

Let me know if you have any questions.
